surreal-engine: init at 2024-11-08

hughobrien opened the pull request:
## Description of changes

Introduce [SurrealEngine](https://github.com/dpjudas/SurrealEngine) 

```
The goal of this project is to reimplement enough of the original Unreal Engine
to make the Unreal Tournament (UT99) maps playable.
```

Testing this package requires [Unreal Tournament (1999)](https://archive.org/details/Nova_UnrealTournamentUSA) to be available. I used wine to run the installer.
```
nix-shell -p p7zip --command '7z x -out99-img "Unreal Tournament (USA) (Disc 1) (Game).7z"'
nix-shell -p bchunk --command 'bchunk ut99-img/*.bin ut99-img/*.cue ut99-img/ut99'
nix-shell -p p7zip --command '7z x -out99-fs ut99-img/ut9901.iso'
WINEPREFIX=${PWD}/ut99-wine nix-shell -p wineWowPackages.full --command "wine ut99-fs/Setup.exe"
```

One should also [patch](https://unrealarchive.org/unreal-tournament/patches-updates/patches/patch-436/index.html) the installation to the version best supported by SurrealEngine
```
WINEPREFIX=${PWD}/ut99-wine nix-shell -p wineWowPackages.full --command "wine UTPatch436nodelta.exe"
```

Then run with
```
nix-shell -I nixpkgs=/data/nixpkgs -p surreal-engine --command "SurrealEngine ut99-wine/drive_c/UnrealTournament"
```

![2024-08-24T14:59:21-04:00](https://github.com/user-attachments/assets/ee6ed60d-f66f-4498-b1cb-584e71457c8e)

If the dialog box is blank, move your cursor around the bottom area where the buttons should be.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

hughobrien commented:
Re-opening this by request (closed before as it was getting stale)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/183

OPNA2608 commented on pkgs/by-name/su/surreal-engine/package.nix:
```diff
@@ -0,0 +1,71 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "surreal-engine";
+  version = "2024-08-31";
```
https://github.com/NixOS/nixpkgs/blob/51031f223c0f05491ae3e24d771ebc76821b6ae1/pkgs/README.md#versioning

```suggestion
  version = "0-unstable-2024-08-31";
```

OPNA2608 commented on pkgs/by-name/su/surreal-engine/package.nix:
```diff
@@ -0,0 +1,71 @@
+  ];
+
+  cmakeBuildType = "Release";
+  parallelBuild = true;
```
I don't think `parallelBuild` is a thing in regular `stdenv.mkDerivation`. Did you mean:

```suggestion
  enableParallelBuilding = true;
```

? That gets auto-enabled by using `cmake`, so this is redundant.

```suggestion
```

OPNA2608 commented on pkgs/by-name/su/surreal-engine/package.nix:
```diff
@@ -0,0 +1,71 @@
+    waylandpp
+  ];
+
+  cmakeBuildType = "Release";
```
Default, redundant.

```suggestion
```

OPNA2608 commented on pkgs/by-name/su/surreal-engine/package.nix:
```diff
@@ -0,0 +1,71 @@
+    cmake
+    makeWrapper
+    pkg-config
+  ];
+  buildInputs = [
```
```suggestion
  ];

  buildInputs = [
```

OPNA2608 reviewed (changes requested):
(I don't know why I was elected to be a reviewer on this, but here goes)

hughobrien commented:
@OPNA2608 thank you for your review, I'm implemented all your suggestions.
Forgive my tagging but since we collaborated on [dosbox-x](https://github.com/NixOS/nixpkgs/pull/258301) I thought you might have an interest in retro games.

OPNA2608 reviewed (changes requested):
Interest yes, time *ehhhh*. Will try.

Please fetch & apply https://github.com/dpjudas/SurrealEngine/pull/142 as a `patches` entry, or bump if it gets merged before you get to it, to fix building on aarch64-linux (haven't tested running yet).

Also, during `configurePhase`:

```
-- Checking for modules 'wayland-client++;wayland-client-extra++;wayland-client-unstable++;wayland-cursor++;xkbcommon>=0.5.0'
--   No package 'xkbcommon' found
```

Needs `libxkbcommon` in `buildInputs` to be happy with that search. Continues then with

```
-- Checking for modules 'wayland-client++;wayland-client-extra++;wayland-client-unstable++;wayland-cursor++;xkbcommon>=0.5.0'
--   Found wayland-client++, version 1.0.0
--   Found wayland-client-extra++, version 1.0.0
--   Found wayland-client-unstable++, version 1.0.0
--   Found wayland-cursor++, version 1.0.0
--   Found xkbcommon, version 1.7.0
Package wayland-client was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-client.pc'
to the PKG_CONFIG_PATH environment variable
Package 'wayland-client', required by 'wayland-client++', not found
Package wayland-client was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-client.pc'
to the PKG_CONFIG_PATH environment variable
Package 'wayland-client', required by 'wayland-client++', not found
Package wayland-client was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-client.pc'
to the PKG_CONFIG_PATH environment variable
Package 'wayland-client', required by 'wayland-client++', not found
Package wayland-client was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-client.pc'
to the PKG_CONFIG_PATH environment variable
Package 'wayland-client', required by 'wayland-client++', not found
```

Haven't checked if it's 100% happy with that arrangement already, otherwise the following would need to be added to `buildInputs` as well: `wayland libffi`

hughobrien commented:
> Please fetch & apply [dpjudas/SurrealEngine#142](https://github.com/dpjudas/SurrealEngine/pull/142) as a `patches` entry, or bump if it gets merged before you get to it, to fix building on aarch64-linux (haven't tested running yet).

Upstream has merged and I've updated the src.

> Needs `libxkbcommon` in `buildInputs` to be happy with that search. Continues then with

Added

>  the following would need to be added to `buildInputs` as well: `wayland libffi`

Added

Thank you for your detailed attention.

hughobrien commented:
Try with gamescope
```
gamescope --expose-wayland --backend sdl -- SurrealEngine /data/ut99-wine/drive_c/UnrealTournament
```

OPNA2608 commented:
> Try with gamescope
> 
> ```
> gamescope --expose-wayland --backend sdl -- SurrealEngine /data/ut99-wine/drive_c/UnrealTournament
> ```

```
↪ gamescope --expose-wayland --backend sdl -- /nix/store/k4kaml0a4v1m5w1nfbzhm9pd5pb8wi11-surreal-engine-0-unstable-2024-11-04/bin/SurrealEngine ~/Games/UnrealTournament
[gamescope] [Info]  console: gamescope version 3.15.11 (gcc 13.3.0)
No CAP_SYS_NICE, falling back to regular-priority compute and threads.
Performance will be affected.
Your Wayland compositor does NOT support wp_presentation/presentation-time which is required for VK_KHR_present_wait and VK_KHR_present_id.
Please complain to your compositor vendor for support. Falling back to X11 window with less accurate present wait.
[gamescope] [Info]  vulkan: selecting physical device 'llvmpipe (LLVM 18.1.8, 128 bits)': queue family 0 (general queue family 0)
[gamescope] [Info]  vulkan: physical device does not support DRM format modifiers
[gamescope] [Error] vulkan: physical device doesn't support VK_EXT_physical_device_drm
terminate called without an active exception
fish: Job 1, 'gamescope --expose-wayland --ba…' terminated by signal SIGABRT (Abbruch)
```

![20241105_18h31m39s_grim](https://github.com/user-attachments/assets/23757624-3c5d-433a-a24d-c74d3da2e65c)

Let's just leave this at "system/setup is struggling, but in theory it *prolly* works fine on ARM" :sweat_smile:.

I'll give it a shot on x86_64 tonight, so we can push this PR along.

LupertEverett commented:
> The initial dialogue window was a solid colour when using SDL's wayland backend, but pressing enter works 

Try moving the mouse around, hovering over the buttons should "force a repaint" or something.

I'm not sure why that happens honestly.

OPNA2608 commented:
`SIGABRT`s when running on non-Wayland after building with Wayland support.

```
Starting program: /nix/store/xsgrhxfdm0xby8zgrjq0zwqdg3g0fzy0-surreal-engine-0-unstable-2024-11-04/bin/.SurrealEngine-wrapped /home/puna/Games/UnrealTournament
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libthread_db.so.1".
terminate called after throwing an instance of 'std::runtime_error'
  what():  Could not connect to Wayland display server via name

Program received signal SIGABRT, Aborted.
0x00007ffff71b2a9c in __pthread_kill_implementation () from /nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libc.so.6
(gdb) bt
#0  0x00007ffff71b2a9c in __pthread_kill_implementation () from /nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libc.so.6
#1  0x00007ffff7160576 in raise () from /nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libc.so.6
#2  0x00007ffff7148935 in abort () from /nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libc.so.6
#3  0x00007ffff74acc2b in __gnu_cxx::__verbose_terminate_handler() [clone .cold] () from /nix/store/s94fwp43xhzkvw8l8nqslskib99yifzi-gcc-13.3.0-lib/lib/libstdc++.so.6
#4  0x00007ffff74bc20a in __cxxabiv1::__terminate(void (*)()) () from /nix/store/s94fwp43xhzkvw8l8nqslskib99yifzi-gcc-13.3.0-lib/lib/libstdc++.so.6
#5  0x00007ffff74bc275 in std::terminate() () from /nix/store/s94fwp43xhzkvw8l8nqslskib99yifzi-gcc-13.3.0-lib/lib/libstdc++.so.6
#6  0x00007ffff74bc4c7 in __cxa_throw () from /nix/store/s94fwp43xhzkvw8l8nqslskib99yifzi-gcc-13.3.0-lib/lib/libstdc++.so.6
#7  0x00007ffff77784f0 in wayland::display_t::display_t(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [clone .cold] () from /nix/store/rdyy1sphkn7mavllg0bswm8w1c03fpz2-waylandpp-1.0.0-lib/lib/libwayland-client++.so.1
#8  0x000000000044167e in __static_initialization_and_destruction_0 () at /nix/store/v7dbfh5n7az2lcap0z1cv4jq0bikya8p-gcc-13.3.0/include/c++/13.3.0/bits/char_traits.h:358
#9  _GLOBAL__sub_I_wayland_display_window.cpp(void) () at /build/source/Thirdparty/ZWidget/src/window/wayland/wayland_display_window.cpp:1271
#10 0x00007ffff714a3ae in __libc_start_main_impl () from /nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libc.so.6
#11 0x00000000004417d5 in _start ()
```

If windowing system selection is done at compile-time and this is expected for now(?), then we should control Wayland support via a switch in the derivation.

OPNA2608 commented:
> Try moving the mouse around, hovering over the buttons should "force a repaint" or something.

That makes it appear yeah.

![20241106_01h59m08s_grim](https://github.com/user-attachments/assets/bc67ca57-2636-48c0-b852-e09a298fefc0)

Note that it paints outside of the window frame here, dunno if that's possibly an issue with the compositor.

LupertEverett commented:
> If windowing system selection is done at compile-time and this is expected for now(?), then we should control Wayland support via a switch in the derivation.

Windowing system selection is controllable with the `ZWIDGET_DISPLAY_BACKEND` environment variable. ZWidget here being the in-house GUI toolkit Surreal Engine uses.

`ZWIDGET_DISPLAY_BACKEND=[Wayland/X11/SDL2]`

Note that Surreal Engine main window (not the launcher) will force the SDL2 backend regardless of the option picked.

> Note that it paints outside of the window frame here, dunno if that's possibly an issue with the compositor.

Does your compositor support server side decorations? ZWidget tries to use em when available. I haven't really tested it without SSDs.

I'll admit that, as the guy who wrote the code for it, the Wayland backend is pretty buggy atm, mostly due to the lack of knowledge and experience on my part.

OPNA2608 commented:
> > If windowing system selection is done at compile-time and this is expected for now(?), then we should control Wayland support via a switch in the derivation.
> 
> Windowing system selection is controllable with the `ZWIDGET_DISPLAY_BACKEND` environment variable. ZWidget here being the in-house GUI toolkit Surreal Engine uses.
> 
> `ZWIDGET_DISPLAY_BACKEND=[Wayland/X11/SDL2]`

```
↪ env ZWIDGET_DISPLAY_BACKEND=SDL2 SurrealEngine ~/Games/UnrealTournament
terminate called after throwing an instance of 'std::runtime_error'
  what():  Could not connect to Wayland display server via name
fish: Job 1, 'env ZWIDGET_DISPLAY_BACKEND=SDL…' terminated by signal SIGABRT (Abbruch)
```

Based on the stack trace from earlier
```
[...]
#6  0x00007ffff74bc4c7 in __cxa_throw () from /nix/store/s94fwp43xhzkvw8l8nqslskib99yifzi-gcc-13.3.0-lib/lib/libstdc++.so.6
#7  0x00007ffff77784f0 in wayland::display_t::display_t(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [clone .cold] () from /nix/store/rdyy1sphkn7mavllg0bswm8w1c03fpz2-waylandpp-1.0.0-lib/lib/libwayland-client++.so.1
#8  0x000000000044167e in __static_initialization_and_destruction_0 () at /nix/store/v7dbfh5n7az2lcap0z1cv4jq0bikya8p-gcc-13.3.0/include/c++/13.3.0/bits/char_traits.h:358
#9  _GLOBAL__sub_I_wayland_display_window.cpp(void) () at /build/source/Thirdparty/ZWidget/src/window/wayland/wayland_display_window.cpp:1271
#10 0x00007ffff714a3ae in __libc_start_main_impl () from /nix/store/3bvxjkkmwlymr0fssczhgi39c3aj1l7i-glibc-2.40-36/lib/libc.so.6
#11 0x00000000004417d5 in _start ()
```
…I think the issue is [the global `WaylandDisplayWindow::s_waylandDisplay` variable at the start of `/Thirdparty/ZWidget/src/window/wayland/wayland_display_window.cpp`](https://github.com/dpjudas/SurrealEngine/blob/7c16d0973307f501fe6765300bfff354d881a7b7/Thirdparty/ZWidget/src/window/wayland/wayland_display_window.cpp#L7), since that always requires running `wayland::display_t()` on startup no matter what backend gets selected later.

LupertEverett commented:
> …I think the issue is [the global WaylandDisplayWindow::s_waylandDisplay variable at the start of /Thirdparty/ZWidget/src/window/wayland/wayland_display_window.cpp](https://github.com/dpjudas/SurrealEngine/blob/7c16d0973307f501fe6765300bfff354d881a7b7/Thirdparty/ZWidget/src/window/wayland/wayland_display_window.cpp#L7), since that always requires running wayland::display_t() on startup no matter what backend gets selected later.

Can you give it a try again when you have time, it should be fixed now.

OPNA2608 commented:
Can confirm that on https://github.com/dpjudas/SurrealEngine/commit/087fa2af7fd0ce51702dd4024b4ae15a88222678 it now launches fine on X11. @hughobrien can you update the PR?

hughobrien committed:
surreal-engine: init at 2024-11-08

OPNA2608 commented:
(Commit title should've been `surreal-engine: init at 0-unstable-2024-11-08`, I've just gone ahead & fixed this in the merge to avoid another back n forth)
